File: oop/playing_cards.py
from enum import Enum
from typing import List
Suit = Enum('Suit', 'clubs diamonds hearts spades')
Values = Enum('Values', 'A ' + ' '.join([str(i) for i in range(2, 11)]) + ' J Q K')
class Card:
    SUITS = [s.name for s in Suit]
    VALUES = [v.name for v in Values]

    def __init__(self, suit, value):
        if suit not in self.SUITS:
            raise ValueError(suit + " not in suits")

        if value not in self.VALUES:
            raise ValueError(value + " not in card values")

        self.__suit = Suit[suit]
        self.__value = Values[value]

    @property
    def value(self) -> int:
        return self.__value.value

    def __str__(self) -> str:
        return f"{self.__value.name} of {self.__suit.name}"

class Deck:
    def __init__(self):
        self.__cards = [Card(suit.name, value.name) for suit in Suit
                                                    for value in Values]

    def find(self, suit, value):
        card = Card(suit, value)
        return self.__cards.index(card)

    def __len__(self):
        return len(self.__cards)

    def __getitem__(self, position):
        return self.__cards[position]

    def __str__(self):
        return str([str(c) for c in self.__cards])

class Game:
    def __init__(self):
        self.deck = Deck()

    def card_beats(self, card_a: int, card_b: int) -> bool:
        return self.deck[card_a].value > self.deck[card_b].value

if __name__ == "__main__":
    game = Game()
    #print(game.deck.find("diamonds", "A"))
    print(game.deck[5])
    print(game.deck[10])
    print("true" if game.card_beats(5, 10) else "false")
    print(Card('hearts', '5') in game.deck)
    for card in reversed(game.deck):
        print(card)

File: oop/ruby/playing_cards.rb
class Suit 
	TYPES = [ "hearts", "diamonds", "spades", "clubs"]
	def initialize(type)
		if !TYPES.include?(type)
			raise ArgumentError.new "#{type} not in #{TYPES}"
		end
		@type = type
	end

	def to_s
		@type
	end
	
	def self.TYPES
		TYPES
	end
end

class Card

	@@values = {"A" => 1, "J" => 11, "Q" => 12, "K" => 13, "Joker" => 14}
	9.times do |x| 
		val = x + 2 
		@@values[val] = val
	end

	def initialize(suit, value)
		@suit = Suit.new(suit)
		if !@@values.has_key?(value)
			raise ArgumentError.new "Invalid value, please use #{@@values.keys}"
		end
		@value = value
	end

	def to_s
		"#{@value} of #{@suit}"
	end

	def value
		@@values[@value]
	end

	def self.values
		@@values
	end
end

class JokerSuit < Suit
	TYPES = ["black", "red"]
	def initialize(suit)
		super(suit)
	end
end

class Joker < Card
	
	def initialize(suit)
		@suit = JokerSuit.new(suit)
		@value = "Joker"
	end
end

class Hand
	attr_accessor :cards
	def initialize
		@cards = []
	end

	def to_s
		@cards.each(&:to_s)
	end
end

class Game
	def initialize
		@cards = []
		@hands = []
	end

	def add_card(suit, value)
		@cards.push(Card.new(suit, value))
	end

	def add_joker(color)
		@cards.push(Joker.new(color))
	end

	def card_string(position)
		@cards[position].to_s
	end

	def add_hand()
		hand = Hand.new
		@hands.push(hand)
		5.times do |x|
			card = Card.new(Suit.TYPES.sample, Card.values.keys.sample)
			@cards.push(card)
			hand.cards.push(card)
		end
	end

	def hand_string(hand)
		@hands[hand].to_s
	end

	def card_beats(a, b)
		@cards[a].value > @cards[b].value
	end
end


# game = Game.new
# game.add_card("hearts", "Q")
# game.add_card("spades", "K")
# game.add_hand
# puts game.hand_string(0)
